chainer/functions/pooling/roi_max_pooling_2d.py

```diff
@@ -0,0 +1,373 @@
+            int wend = static_cast<int>(ceil(static_cast<float>(pw + 1)
+                                        * bin_size_w));
+
+            // Add roi offsets and clip to input boundaries
+            hstart = min(max(hstart + roi_start_h, 0), height);
+            hend = min(max(hend + roi_start_h, 0), height);
+            wstart = min(max(wstart + roi_start_w, 0), width);
+            wend = min(max(wend + roi_start_w, 0), width);
+            bool is_empty = (hend <= hstart) || (wend <= wstart);
+
+            // Define an empty pooling region to be zero
+            float maxval = is_empty ? 0 : -1E+37;
+            // If nothing is pooled, argmax=-1 causes nothing to be backprop'd
+            int maxidx = -1;
+            int data_offset = (roi_batch_ind * channels + c) * height * width;
+            for (int h = hstart; h < hend; ++h) {
+                for (int w = wstart; w < wend; ++w) {
+                    int bottom_index = h * width + w;
+                    if (bottom_data[data_offset + bottom_index] > maxval) {
+                        maxval = bottom_data[data_offset + bottom_index];
+                        maxidx = bottom_index;
+                    }
+                }
+            }
+            top_data = maxval;
+            argmax_data = maxidx;
+            ''', 'roi_max_pooling_2d_fwd'
+        )(bottom_data, bottom_rois, bottom_roi_indices,
+          self.spatial_scale, channels, height, width,
+          self.outh, self.outw, top_data, self.argmax_data)
+
+        return top_data,
+
+    def backward_cpu(self, inputs, gy):
+        bottom_rois, bottom_roi_indices = inputs[1:]
+        channels, height, width = self._bottom_data_shape[1:]
+        n_rois = bottom_rois.shape[0]
+        bottom_delta = numpy.zeros(self._bottom_data_shape, bottom_rois.dtype)
+
+        for i_roi in six.moves.range(n_rois):
+            idx = bottom_roi_indices[i_roi]
+            ymin, xmin, ymax, xmax = bottom_rois[i_roi]
+            ymin = int(round(ymin * self.spatial_scale))
+            xmin = int(round(xmin * self.spatial_scale))
+            ymax = int(round(ymax * self.spatial_scale))
+            xmax = int(round(xmax * self.spatial_scale))
+            roi_height = max(ymax - ymin, 1)
+            roi_width = max(xmax - xmin, 1)
+
+            strideh = float(roi_height) / float(self.outh)
+            stridew = float(roi_width) / float(self.outw)
+
+            # iterate all the w, h (from feature map) that fall into this ROIs
+            for w in six.moves.range(xmin, xmax + 1):
+                for h in six.moves.range(ymin, ymax + 1):
+                    phstart = int(numpy.floor(float(h - ymin) / strideh))
+                    phend = int(numpy.ceil(float(h - ymin + 1) / strideh))
+                    pwstart = int(numpy.floor(float(w - xmin) / stridew))
+                    pwend = int(numpy.ceil(float(w - xmin + 1) / stridew))
+
+                    phstart = min(max(phstart, 0), self.outh)
```

Comment:
Do we need to compare `phstart` with 0?

Reply:
I just followed https://github.com/chainer/chainer/blob/master/chainer/functions/pooling/roi_pooling_2d.py.
IMO, we do not need to compare, but there might be some exception.
Also, I looked into Caffe2, and they have completely different but simple kernel.
Should we move to this simple one?
https://github.com/pytorch/pytorch/blob/master/caffe2/operators/roi_pool_op.cu#L101-L119

Reply:
I updated to follow caffe2 kernel.